service-tests/pypi.js

```diff
@@ -0,0 +1,222 @@
+'use strict';
+
+const Joi = require('joi');
+const ServiceTester = require('./runner/service-tester');
+const {
+  isCommaSeperatedPythonVersions,
+  isSemver,
+} = require('./helpers/validators');
+
+const t = new ServiceTester({ id: 'pypi', title: 'PyPi badges' });
+module.exports = t;
+
+
+/*
+  Note:
+  Download statistics are no longer available from pypi
+  it is exptected that the download badges all show
+  'no longer available'
+*/
+t.create('daily downloads (expected failure)')
+  .get('/dd/djangorestframework.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'downloads',
+    value: 'no longer available'
+  }));
+
+t.create('weekly downloads (expected failure)')
+  .get('/dw/djangorestframework.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'downloads',
+    value: 'no longer available'
+  }));
+
+t.create('monthly downloads (expected failure)')
+  .get('/dm/djangorestframework.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'downloads',
+    value: 'no longer available'
+  }));
+
+t.create('daily downloads (invalid)')
+  .get('/dd/not-a-package.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'pypi',
+    value: 'invalid'
+  }));
+
+t.create('weekly downloads (invalid)')
+  .get('/dw/not-a-package.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'pypi',
+    value: 'invalid'
+  }));
+
+t.create('monthly downloads (invalid)')
+  .get('/dm/not-a-package.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'pypi',
+    value: 'invalid'
+  }));
+
+/*
+  Note:
+  Not all project on PyPi follow SemVer
+
+  Versions strings like
+  - 2.7.3.2
+  - 2.0rc1
+  - 0.1.30b10
+  are prefectly legal.
+
+  We'll run this test againt a project that follows SemVer...
+*/
+t.create('version (semver)')
+  .get('/v/requests.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'pypi',
+    value: isSemver
+  }));
+
+/*
+  ..whereas this project uses version numbers that
+  should just be validated as an arbitary string
+*/
```

Comment:
I appreciate that you're wanting to make a robust test, though I do think it's important to use something more restrictive than `Joi.string()`, which will silently pass with strings like `''`, `'undefined'` or `'invalid'`.

```console
❯ pip install psycopg2==x
Collecting psycopg2==x
  Could not find a version that satisfies the requirement psycopg2==x (from versions: 2.0.10, 2.0.11, 2.0.12, 2.0.13, 2.0.14, 2.2.0, 2.2.1, 2.2.2, 2.3.0, 2.3.1, 2.3.2, 2.4, 2.4.1, 2.4.2, 2.4.3, 2.4.4, 2.4.5, 2.4.6, 2.5, 2.5.1, 2.5.2, 2.5.3, 2.5.4, 2.5.5, 2.6, 2.6.1, 2.6.2, 2.7, 2.7.1, 2.7.2, 2.7.3, 2.7.3.1, 2.7.3.2)
No matching distribution found for psycopg2==x
```

There's a regex in [`validators.js`](https://github.com/badges/shields/blob/master/service-tests/helpers/validators.js) that should work, `isVPlusDottedVersionAtLeastOne`.